### PR TITLE
Coin V5 Upgrades + Documented Process

### PR DESCRIPTION
**Pull request by emilypi**

### CODE REVIEW

**Review by larskuhtz — APPROVED** [review]

File: docs/CoinContractUpgrades.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+- PactInProcApi
+  - Make sure v* is backcompatible across 2 forks so (see: pact4coin3UpgradeTest)
+    - https://github.com/kadena-io/chainweb-node/pull/1416
+    - https://github.com/kadena-io/chainweb-node/pull/1413
+
+  - It should be the case that updates are backcompatible if users start from scratch
+  - Events need to be backcompatible
+
+- `test/Chainweb/Test/Pact/ModuleCacheOnRestart.hs`
```
[the remark this answers is not in the document]
Comment (emilypi) [reply docs/CoinContractUpgrades.md:154]: Will do

**Comment by emilypi** [comment]
interesting. I'll sort out why.

File: docs/CoinContractUpgrades.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+      , coinContractV4
+      ]
+
+    genCoinV5Payloads :: IO ()
+    genCoinV5Payloads = genTxModule "CoinV5"
+      [ coinContractV5
+      ]
+  ```
+  3. the newly updated `genCoinV*Payloads` function must be added to the `main` function. Keep in mind that if any additional upgrades are present, the transactions are generated in sequence, therefore are order-sensitive. If new code has dependencies, then those dependencies need to occur earlier in the input list to `genTxModule`.
+  4. The `ea` tool must be run by issuing the following command:
```
Comment (jmcardon) [inline docs/CoinContractUpgrades.md:52]: alternatively IIRC you can just go into the Ea in the repl and run main IIRC. :^)

Useful if you're already in the repl session fixing something.

File: docs/CoinContractUpgrades.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+-- | Pact and coin contract changes for Chainweb 2.14
+--
+chainweb214Pact
+    :: AtOrAfter
+    -> ChainwebVersion
+    -> BlockHeight
+    -> Bool
+chainweb214Pact aoa v h = case aoa of
+    At -> go (==) v h
+    After -> go (flip (>)) v h
```
Comment (jmcardon) [inline docs/CoinContractUpgrades.md:135]: `(<=)` over `flip (>)`

File: docs/CoinContractUpgrades.md (hunk 2 above)
Comment (emilypi) [reply docs/CoinContractUpgrades.md:52]: Done :)

File: src/Chainweb/Version.hs [hunk 4]
```diff
@@ -953,8 +955,8 @@ chainweb215Pact aoa v h = case aoa of
     At -> go (==) v h
     After -> go (<) v h
   where
-    go f Mainnet01 = f 2766637 --  2022-06-17 00:00:00+00:00
-    go f Testnet04 = f 2295440 --  2022-06-16 12:00:00+00:00
+    go f Mainnet01 = f 2766637 -- 2022-06-17T00:00:00+00:00
+    go f Testnet04 = f 2295440 -- 2022-06-16T12:00:00+00:00
```
Comment (larskuhtz) [inline src/Chainweb/Version.hs:959]: Most recent estimates:

|    | network   | time                      |   height |
|---:|:----------|:--------------------------|---------:|
|  0 | testnet   | 2022-06-16 12:00:00+00:00 |  2295437 |
|  1 | mainnet   | 2022-06-17 00:00:00+00:00 |  2766630 |

```suggestion
    go f Mainnet01 = f 2766630 -- 2022-06-17T00:00:00+00:00
    go f Testnet04 = f 2295437 -- 2022-06-16T12:00:00+00:00
```

File: test/Chainweb/Test/Pact/PactInProcApi.hs [hunk 5]
```diff
@@ -582,6 +583,62 @@ pact43UpgradeTest bdb mpRefIO pact = do
...
+  -- execute pre-fork xchain transfer (blocc0)
+  setOneShotMempool mpRefIO blocc0
+  runCut' -- 34
+  pwo34 <- getPWO bdb cid
+  tx34_0 <- txResult "pwo34_0" 0 pwo34
+  tx34_1 <- txResult "pwo34_1" 1 pwo34
+  assertSatisfies "tx34_0 success" (_pactResult $ _crResult tx34_0) isRight
+  assertSatisfies "tx34_1 success" (_pactResult $ _crResult tx34_1) isRight
+
+  putStrLn "HERE"
```
Comment (larskuhtz) [inline test/Chainweb/Test/Pact/PactInProcApi.hs:601]: Is this supposed to be HERE?

**Review by larskuhtz — APPROVED** [review]
Approved. I didn't review the pact code and only quickly skimmed over the tests. Someone else should take a look at that.

File: test/Chainweb/Test/Pact/PactInProcApi.hs (hunk 5 above)
Comment (emilypi) [reply test/Chainweb/Test/Pact/PactInProcApi.hs:601]: Still working out these tests, so yes for the time being :)

File: src/Chainweb/Pact/TransactionExec.hs [hunk 6]
```diff
@@ -507,6 +508,7 @@ applyUpgrades v cid height
     applyCoinV3 = applyTxs coinV3Transactions [FlagDisableInlineMemCheck, FlagDisablePact43]
 
     applyCoinV4 = applyTxs coinV4Transactions []
```
Comment (jmcardon) [inline src/Chainweb/Pact/TransactionExec.hs:510]: this latest flag is mostly backcompat in pact so it's nice.

File: test/Chainweb/Test/Pact/PactInProcApi.hs [hunk 7]
```diff
@@ -582,6 +584,86 @@ pact43UpgradeTest bdb mpRefIO pact = do
         $ mkKeySetData "k" [sender00]
     runCut' = runCut testVersion bdb pact (offsetBlockTime second) zeroNoncer noMiner
 
+chainweb215Test :: TestBlockDb -> IO (IORef MemPoolAccess) -> WebPactExecutionService -> IO ()
+chainweb215Test bdb mpRefIO pact = do
+
+  -- run past genesis, upgrades
+  runCutN' 30 -- 1->30
```
Comment (jmcardon) [inline test/Chainweb/Test/Pact/PactInProcApi.hs:591]: BASED

File: test/Chainweb/Test/Pact/PactInProcApi.hs [hunk 8]
```diff
@@ -582,6 +584,86 @@ pact43UpgradeTest bdb mpRefIO pact = do
...
+  -- - transfer
+  -- - x_resume
+
+  cbev1 <- mkTransferEvent "sender00" "NoMiner" 0.0264 "coin" v5Hash
+  recdev1 <- mkTransferXChainRecdEvent "" "sender00" 0.0123 "coin" v5Hash "8"
+  tev1 <- mkTransferEvent "" "sender00" 0.0123 "coin" v5Hash
+  xev1 <- mkXResumeEvent "sender00" "sender00" 0.0123 sender00Ks "pact" v4Hash "8" "0"
+  assertEqual "Transfer events @ block 42" [cbev1, tev1, recdev1, xev1] $ _crEvents tx42_0
+  assertSatisfies "tx42_0 success" (_pactResult $ _crResult tx42_0) isRight
+
```
Comment (jmcardon) [inline test/Chainweb/Test/Pact/PactInProcApi.hs:633]: Just for documenting what we already discussed: test here for the `TRANSFER_XCHAIN_RECD` event and x-chain transfer for coinv5.

**Pull request merged**